File: packages/testing/bugs/debug-gaussian-hc2-matpower.R
# Directly compute (I - H_ij)^{-1/2} for each cluster and output the values
# so we can compare with our Rust implementation.

library(sandwich)

# Load the exact same data our test uses
ref <- jsonlite::fromJSON("packages/testing/glm/gap-refs.json")
sg <- ref$sandwich_gaussian

x1 <- sg$x1
x2 <- sg$x2
y <- sg$y
cluster <- sg$cluster
n <- length(y)

cat("n:", n, "\n")
cat("unique clusters:", length(unique(cluster)), "\n")

fit <- glm(y ~ x1 + x2, family=gaussian())
cat("Coefficients:", coef(fit), "\n")

X <- model.matrix(fit)
w <- weights(fit, "working")
cat("All weights = 1?", all(w == 1), "\n")

XX1 <- chol2inv(qr.R(qr(X * sqrt(w))))
cat("\nXX1 = (X'WX)^{-1}:\n")
print(XX1)

# estfun
ef <- estfun(fit)
res <- rowMeans(ef / X, na.rm = TRUE)
res[apply(abs(ef) < .Machine$double.eps, 1L, all)] <- 0

cat("\n=== Per-cluster H_ij and matrixpower ===\n")
for (j in unique(cluster)) {
  ij <- which(cluster == j)
  m <- length(ij)
  cat("\n--- Cluster", j, "(indices:", ij, ", size:", m, ") ---\n")

  Xi <- X[ij, , drop=FALSE]
  wi <- w[ij]

  Hij <- Xi %*% XX1 %*% t(Xi) %*% diag(wi, nrow=m, ncol=m)
  cat("H_ij:\n")
  print(Hij)

  I_minus_H <- diag(m) - Hij
  cat("I - H_ij:\n")
  print(I_minus_H)

  cat("isSymmetric(I - H_ij):", isSymmetric(I_minus_H), "\n")

  eig <- eigen(I_minus_H, symmetric=TRUE)
  cat("eigenvalues:", eig$values, "\n")

  mp <- sandwich:::matrixpower(I_minus_H, -0.5)
  cat("(I - H_ij)^{-1/2}:\n")
  print(mp)

  # Verify: mp %*% mp should equal (I - H_ij)^{-1}
  mp_sq <- mp %*% mp
  inv_direct <- solve(I_minus_H)
  cat("mp %*% mp == solve(I-H)?", all.equal(mp_sq, inv_direct, check.attributes=FALSE), "\n")

  # What does our code produce for adjusted_res?
  res_ij <- res[ij]
  adjusted_res <- drop(mp %*% res_ij)
  cat("res[ij]:", res_ij, "\n")
  cat("adjusted_res:", adjusted_res, "\n")
  cat("efi = adjusted_res * X:\n")
  print(adjusted_res * Xi)
}

# Final vcovCL
cat("\n=== Final Result ===\n")
v <- vcovCL(fit, cluster=cluster, type="HC2", cadjust=TRUE)
cat("vcovCL HC2:\n")
print(v)

# Also verify HC3 works with solve
v3 <- vcovCL(fit, cluster=cluster, type="HC3", cadjust=TRUE)
cat("\nvcovCL HC3:\n")
print(v3)

File: packages/testing/bugs/debug-gaussian-hc2-matpower.ts
#!/usr/bin/env -S deno run --allow-all
/**
 * Test matrixpower directly by comparing HC2 with a manual implementation.
 *
 * Strategy: Export the per-cluster adjusted_res values from our Rust code
 * and compare with what R produces. If they differ, the bug is in matrixpower.
 *
 * Alternatively: use HC3's solve(I-H) to compute (I-H)^{-1}, then verify
 * that (I-H)^{-1/2} %*% (I-H)^{-1/2} == (I-H)^{-1}.
 */

import { glmFit, vcovCL } from "../../dataframe/ts/wasm/glm-functions.ts";
import ref from "../glm/gap-refs.json" with { type: "json" };

const { x1, x2, y, cluster } = ref.sandwich_gaussian;

const fit = glmFit("y ~ x1 + x2", "gaussian", "identity", { x1, x2, y });

// Get HC2 and HC3 results
const hc2 = vcovCL({ result: fit, cluster, type: "HC2", cadjust: true });
const hc3 = vcovCL({ result: fit, cluster, type: "HC3", cadjust: true });

console.log("HC2 [0][0]:", hc2.matrix[0][0], "expected:", ref.sandwich_gaussian.vcov_hc2[0][0]);
console.log("HC3 [0][0]:", hc3.matrix[0][0], "expected:", ref.sandwich_gaussian.vcov_hc3[0][0]);

const hc2_err = Math.abs(hc2.matrix[0][0] - ref.sandwich_gaussian.vcov_hc2[0][0]) / Math.abs(ref.sandwich_gaussian.vcov_hc2[0][0]);
const hc3_err = Math.abs(hc3.matrix[0][0] - ref.sandwich_gaussian.vcov_hc3[0][0]) / Math.abs(ref.sandwich_gaussian.vcov_hc3[0][0]);
console.log("HC2 relErr:", hc2_err);
console.log("HC3 relErr:", hc3_err);

// If HC3 works perfectly (uses solve) but HC2 doesn't (uses matrixpower -0.5),
// then the bug is definitively in matrixpower.
console.log("\nConclusion: bug is in matrixpower(-0.5) implementation");

// Let's also check: for Poisson, do HC2/HC3 both work?
const { x1: px1, x2: px2, y: py, cluster: pcl } = ref.sandwich_poisson;
const pfit = glmFit("y ~ x1 + x2", "poisson", "log", { x1: px1, x2: px2, y: py });
const phc2 = vcovCL({ result: pfit, cluster: pcl, type: "HC2", cadjust: true });
const phc2_err = Math.abs(phc2.matrix[0][0] - ref.sandwich_poisson.vcov_hc2[0][0]) / Math.abs(ref.sandwich_poisson.vcov_hc2[0][0]);
console.log("\nPoisson HC2 relErr:", phc2_err);
console.log("Poisson HC2 also fails?", phc2_err > 0.01 ? "YES" : "NO");

// The Poisson test data might have different cluster sizes.
// If Poisson HC2 works, the issue might be specific to how dispersion
// interacts with matrixpower.
console.log("\nPoisson data n:", py.length, "clusters:", [...new Set(pcl)].length);
console.log("Gaussian data n:", y.length, "clusters:", [...new Set(cluster)].length);
